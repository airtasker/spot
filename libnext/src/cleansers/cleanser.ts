import { ContractDefinition } from "../models/definitions";
import { ContractNode } from "../models/nodes";
import { cleanseApi } from "./nodes/api-cleanser";
import { cleanseEndpoint } from "./nodes/endpoint-cleanser";

/**
 * Removes all additional parsing information associated with a parsed contract.
 *
 * @param contractNode a parsed contract
 */
export function cleanse(contractNode: ContractNode): ContractDefinition {
  const api = cleanseApi(contractNode.api.value);
  const endpoints = contractNode.endpoints.map(endpoint =>
    cleanseEndpoint(endpoint.value)
  );
  const types = contractNode.types;

  return { api, endpoints, types };
}
