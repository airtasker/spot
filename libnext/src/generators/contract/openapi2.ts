import assertNever from "assert-never";
import * as YAML from "js-yaml";
import {
  BodyNode,
  ContractNode,
  DefaultResponseNode,
  EndpointNode
} from "../../models/nodes";
import {
  OpenAPI2SchemaType,
  OpenAPI2SchemaTypeObject,
  openApi2TypeSchema
} from "./openapi2-schema";
import compact = require("lodash/compact");
import pickBy = require("lodash/pickBy");

export function generateOpenApiV2(
  contractNode: ContractNode,
  format: "json" | "yaml"
) {
  const contract = openApiV2(contractNode);
  switch (format) {
    case "json":
      return JSON.stringify(contract, null, 2);
    case "yaml":
      return YAML.safeDump(contract);
    default:
      throw assertNever(format);
  }
}

export function openApiV2(contractNode: ContractNode): OpenApiV2 {
  return {
    swagger: "2.0",
    info: {
      version: "0.0.0",
      title: contractNode.api.name,
      ...pickBy({ description: contractNode.api.description }),
      contact: {
        name: "TODO"
      }
    },
    paths: contractNode.endpoints.reduce(
      (acc, endpoint) => {
        const openApiPath = endpoint.path.replace(/:(\w+)/g, "{$1}");
        acc[openApiPath] = acc[openApiPath] || {};
        acc[openApiPath][endpoint.method.toLowerCase()] = {
          operationId: endpoint.name,
          description: endpoint.description,
          tags: endpoint.tags,
          parameters: getParameters(endpoint),
          responses: {
            ...(endpoint.defaultResponse
              ? { default: response(endpoint.defaultResponse) }
              : {}),
            ...endpoint.responses.reduce<{
              [statusCode: string]: OpenAPIV2Response;
            }>((acc, responseNode) => {
              acc[responseNode.status.toString(10)] = response(responseNode);
              return acc;
            }, {})
          }
        };
        return acc;
      },
      {} as {
        [endpointPath: string]: {
          [method: string]: OpenAPIV2Operation;
        };
      }
    ),
    definitions: contractNode.types.reduce<{
      [typeName: string]: OpenAPI2SchemaType;
    }>((acc, typeNode) => {
      acc[typeNode.name] = openApi2TypeSchema(typeNode.type);
      return acc;
    }, {})
  };
}

function getParameters(endpoint: EndpointNode): OpenAPIV2Parameter[] {
  const parameters = endpoint.request.pathParams
    .map(
      (pathParam): OpenAPIV2Parameter => {
        const schemaType = openApi2TypeSchema(pathParam.type);
        if ("type" in schemaType && schemaType.type === "object") {
          throw new Error(`Unsupported object type in path parameter`);
        }
        return {
          in: "path",
          name: pathParam.name,
          description: pathParam.description,
          ...schemaType,
          required: true
        };
      }
    )
    .concat(endpoint.request.body ? [requestBody(endpoint.request.body)] : [])
    .concat(
      endpoint.request.queryParams.map(
        (queryParam): OpenAPIV2Parameter => {
          const schemaType = openApi2TypeSchema(queryParam.type);
          if ("type" in schemaType && schemaType.type === "object") {
            throw new Error(`Unsupported object type in query parameter`);
          }
          return {
            in: "query",
            name: queryParam.name,
            description: queryParam.description,
            ...schemaType,
            required: !queryParam.optional
          };
        }
      )
    )
    .concat(
      endpoint.request.headers.map(
        (header): OpenAPIV2Parameter => {
          const schemaType = openApi2TypeSchema(header.type);
          if ("type" in schemaType && schemaType.type === "object") {
            throw new Error(`Unsupported object type in header`);
          }
          return {
            in: "header",
            name: header.name,
            description: header.description,
            ...schemaType,
            required: !header.optional
          };
        }
      )
    );
  return compact(parameters);
}

function requestBody(body: BodyNode): OpenAPIV2Parameter {
  return {
    in: "body",
    name: "body",
    description: "TODO",
    required: !body.optional,
    schema: openApi2TypeSchema(body.type)
  };
}

function response(response: DefaultResponseNode): OpenAPIV2Response {
  return {
    schema: response.body && openApi2TypeSchema(response.body.type),
    description: response.description || ""
  };
}

export interface OpenApiV2 {
  swagger: "2.0";
  tags?: OpenAPIV2TagObject[];
  info: {
    version: string;
    title: string;
    description?: string;
    termsOfService?: string;
    contact?: {
      name?: string;
      url?: string;
      email?: string;
    };
    license?: {
      name: string;
      url?: string;
    };
  };
  host?: string;
  paths: {
    [endpointPath: string]: {
      [method: string]: OpenAPIV2Operation;
    };
  };
  definitions: {
    [typeName: string]: OpenAPI2SchemaType;
  };
}

export interface OpenAPIV2TagObject {
  name: string;
  description?: string;
}

export interface OpenAPIV2Operation {
  operationId: string;
  description?: string;
  tags?: string[];
  parameters: OpenAPIV2Parameter[];
  requestBody?: OpenAPI2SchemaType;
  responses: {
    default?: OpenAPIV2Response;
    // Note: we use | undefined because otherwise "default" would have to be required.
    [statusCode: string]: OpenAPIV2Response | undefined;
  };
}

export type OpenAPIV2Parameter =
  | OpenAPIV2BodyParameter
  | OpenAPIV2NonBodyParameter;

export type OpenAPIV2NonBodyParameter = {
  in: "path" | "query" | "header";
  name: string;
  description?: string;
  required: boolean;
} & Exclude<OpenAPI2SchemaType, OpenAPI2SchemaTypeObject>;

export type OpenAPIV2BodyParameter = {
  in: "body";
  name: "body";
  description?: string;
  required: boolean;
  schema: OpenAPI2SchemaType | undefined;
};

export interface OpenAPIV2Response {
  schema?: OpenAPI2SchemaType;
  description: string;
}
