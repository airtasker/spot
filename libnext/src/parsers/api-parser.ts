import { ClassDeclaration } from "ts-simple-ast";
import {
  extractJsDocComment,
  extractStringProperty,
  extractDecoratorFactoryConfiguration
} from "./utilities/parser-utility";
import { ApiDefinition } from "../models/definitions";

/**
 * Parse an `@api` decorated class.
 *
 * @param klass a class declaration
 */
export function parseApi(klass: ClassDeclaration): ApiDefinition {
  const decorator = klass.getDecoratorOrThrow("api");
  const description = extractJsDocComment(klass);
  const configuration = extractDecoratorFactoryConfiguration(decorator);
  const name = extractStringProperty(configuration, "name");
  return { name, description };
}
