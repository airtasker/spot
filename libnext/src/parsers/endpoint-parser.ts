import { ClassDeclaration, ObjectLiteralExpression } from "ts-simple-ast";
import {
  extractJsDocComment,
  extractDecoratorFactoryConfiguration,
  extractStringProperty,
  isHttpMethod
} from "./parser-utility";
import { EndpointDefinition } from "../models/definitions";
import { HttpMethod } from "../models/types";

/**
 * Parse an Endpoint definition object.
 *
 * @param klass a class declaration
 */
export function parseEndpoint(klass: ClassDeclaration): EndpointDefinition {
  const decorator = klass.getDecoratorOrThrow("endpoint");
  const description = extractJsDocComment(klass);
  const configuration = extractDecoratorFactoryConfiguration(decorator);
  const method = extractHttpMethodProperty(configuration, "method");
  const path = extractStringProperty(configuration, "path");
  return { description, method, path };
}

/**
 * Extract a HTTP method property from an object literal.
 *
 * @param objectLiteral an object literal
 * @param propertyName the property to extract
 */
function extractHttpMethodProperty(
  objectLiteral: ObjectLiteralExpression,
  propertyName: string
): HttpMethod {
  const method = extractStringProperty(objectLiteral, propertyName);
  if (!isHttpMethod(method)) {
    throw new Error(`expected a HttpMethod, got ${method}`);
  }
  return method;
}
