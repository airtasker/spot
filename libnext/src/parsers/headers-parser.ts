import { HeaderDefinition } from "../models/definitions";
import { ParameterDeclaration } from "ts-simple-ast";
import {
  ensureNodeNotOptional,
  extractObjectParameterProperties,
  ensureDataTypeIsKind,
  extractJsDocComment
} from "./utilities/parser-utility";
import { parseType } from "./utilities/type-parser";
import { KindOfString, KindOfNumber } from "../models/types";

/**
 * Parse an `@headers` decorated parameter.
 *
 * @param parameter a parameter declaration
 */
export function parseHeaders(
  parameter: ParameterDeclaration
): HeaderDefinition[] {
  parameter.getDecoratorOrThrow("headers");
  ensureNodeNotOptional(parameter);
  const properties = extractObjectParameterProperties(parameter);
  return properties.map(property => {
    const propertyDataType = parseType(property.getType());
    ensureDataTypeIsKind(propertyDataType, KindOfString.concat(KindOfNumber));
    return {
      name: property.getName(),
      description: extractJsDocComment(property),
      type: propertyDataType,
      optional: property.hasQuestionToken()
    };
  });
}
