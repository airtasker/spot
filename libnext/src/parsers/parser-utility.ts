import {
  ts,
  ObjectLiteralExpression,
  JSDocableNode,
  Decorator,
  TypeGuards
} from "ts-simple-ast";
import { HttpMethod } from "../../../lib/src/lib";

/**
 * Extracts the JS Doc comment from a node.
 *
 * @param node a JS Docable Node
 */
export function extractJsDocComment(node: JSDocableNode): string | undefined {
  const jsDocs = node.getJsDocs();
  if (jsDocs.length === 1) {
    return jsDocs[0].getComment();
  } else if (jsDocs.length > 1) {
    throw new Error(`expected 1 jsDoc node, got ${jsDocs.length}`);
  }
  return;
}

/**
 * Extract a string property from an object literal.
 *
 * @param objectLiteral an object literal
 * @param propertyName the property to extract
 */
export function extractStringProperty(
  objectLiteral: ObjectLiteralExpression,
  propertyName: string
): string {
  return objectLiteral
    .getPropertyOrThrow(propertyName)
    .getLastChildIfKindOrThrow(ts.SyntaxKind.StringLiteral)
    .getLiteralText();
}

/**
 * Extract an optional string property from an object literal.
 *
 * @param objectLiteral an object literal
 * @param propertyName the property to extract
 */
export function extractOptionalStringProperty(
  objectLiteral: ObjectLiteralExpression,
  propertyName: string
): string | undefined {
  const property = objectLiteral.getProperty(propertyName);
  if (property) {
    return property
      .getLastChildIfKindOrThrow(ts.SyntaxKind.StringLiteral)
      .getLiteralText();
  }
  return;
}

/**
 * Extract the Configuration object (first argument) from a Spot decorator factory.
 *
 * @param decorator a decorator factory
 */
export function extractDecoratorFactoryConfiguration(
  decorator: Decorator
): ObjectLiteralExpression {
  if (decorator.getArguments().length === 1) {
    const argument = decorator.getArguments()[0];
    if (TypeGuards.isObjectLiteralExpression(argument)) {
      return argument;
    } else {
      throw new Error(`expected object literal`);
    }
  } else {
    throw new Error(
      `expected 1 argument, got ${decorator.getArguments().length}`
    );
  }
}

/**
 * Determine if a string is a HTTP method.
 *
 * @param method the string
 */
export function isHttpMethod(method: string): method is HttpMethod {
  switch (method) {
    case "GET":
    case "HEAD":
    case "POST":
    case "PUT":
    case "DELETE":
    case "CONNECT":
    case "OPTIONS":
    case "TRACE":
    case "PATCH":
      return true;
    default:
      return false;
  }
}
