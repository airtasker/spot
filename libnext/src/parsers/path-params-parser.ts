import { ParameterDeclaration, TypeGuards } from "ts-simple-ast";
import {
  extractJsDocComment,
  extractPropertySignature
} from "./utilities/parser-utility";
import { parseType } from "./utilities/type-parser";
import { PathParamDefinition } from "../models/definitions";

/**
 * Parse an `@pathParams` decorated parameter.
 *
 * @param parameter a parameter declaration
 */
export function parsePathParams(
  parameter: ParameterDeclaration
): PathParamDefinition[] {
  // sanity check
  parameter.getDecoratorOrThrow("pathParams");
  if (parameter.hasQuestionToken()) {
    throw new Error("parameter cannot be optional");
  }
  const type = parameter.getType();
  if (!type.isObject()) {
    throw new Error("expected object");
  }
  const properties = type.getProperties();
  return properties.map(property => {
    const propertySignature = extractPropertySignature(property);
    return {
      name: propertySignature.getName(),
      description: extractJsDocComment(propertySignature),
      type: parseType(propertySignature.getType())
    };
  });
}
