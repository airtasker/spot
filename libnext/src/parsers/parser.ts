import { CompilerOptions, Project, SourceFile, ts } from "ts-simple-ast";
import { ContractNode, EndpointNode } from "../models/nodes";
import { ReferenceType } from "../models/types";
import { parseApi } from "./nodes/api-parser";
import { parseEndpoint } from "./nodes/endpoint-parser";
import { extractJsDocComment } from "./utilities/parser-utility";
import { parseInterfaceDeclaration, parseType } from "./utilities/type-parser";
import {
  retrieveTypeReferencesFromEndpoints,
  retrieveTypeReferencesFromType,
  uniqueReferences
} from "./utilities/type-reference-resolver";

export function parseFilePath(
  sourcePath: string,
  customCompilerOptions: CompilerOptions = {}
): ContractNode {
  const defaultCompilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true
  };

  const compilerOptions = {
    ...defaultCompilerOptions,
    ...customCompilerOptions
  };

  // Creates a new typescript program in memory
  const project = new Project({ compilerOptions });
  const sourceFile = project.addExistingSourceFile(sourcePath);

  // Add all dependent files that the project requires
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  return parseRootSourceFile(sourceFile, project);
}

/**
 * Parse a root source file to return a contract.
 */
function parseRootSourceFile(
  file: SourceFile,
  projectContext: Project
): ContractNode {
  const apiClasses = file
    .getClasses()
    .filter(klass => klass.getDecorator("api") !== undefined);

  if (apiClasses.length !== 1) {
    throw new Error(
      `expected class decorated with @api to be defined exactly once in root source file, found ${
        apiClasses.length
      } usages`
    );
  }
  const api = parseApi(apiClasses[0]);

  const sourceEndpoints = file
    .getClasses()
    .filter(klass => klass.getDecorator("endpoint") !== undefined)
    .map(endpointClass => parseEndpoint(endpointClass));

  const importedFiles = file
    .getImportDeclarations()
    .map(myImport => myImport.getModuleSpecifierSourceFileOrThrow());

  const endpoints = importedFiles.reduce<EndpointNode[]>(
    (endpointsAcc, currentFile) =>
      currentFile
        .getClasses()
        .filter(klass => klass.getDecorator("endpoint") !== undefined)
        .map(endpointClass => parseEndpoint(endpointClass))
        .concat(endpointsAcc),
    sourceEndpoints
  );

  // Direct reference types, filtered to unique references for efficiency
  const uniqueDirectReferenceTypes = uniqueReferences(
    retrieveTypeReferencesFromEndpoints(endpoints)
  );

  // Reference types from the direct reference type hierarchy
  const secondaryReferenceTypes = uniqueDirectReferenceTypes.reduce<
    ReferenceType[]
  >(
    (referenceTypesAcc, currentReferenceType) =>
      referenceTypesAcc.concat(
        retrieveTypeReferencesFromType(currentReferenceType, projectContext)
      ),
    []
  );

  // Combine and filter to unique type references
  const allReferenceTypes = uniqueReferences(
    uniqueDirectReferenceTypes.concat(secondaryReferenceTypes)
  );

  // Construct the equivalent TypeNode for each reference type
  const types = allReferenceTypes.map(referenceType => {
    const file = projectContext.getSourceFileOrThrow(referenceType.location);

    const typeAlias = file.getTypeAlias(referenceType.name);
    if (typeAlias !== undefined) {
      const name = typeAlias.getName();
      const description = extractJsDocComment(typeAlias);
      const type = parseType(typeAlias.getTypeNodeOrThrow());
      return { description, name, type };
    }

    const interfaceDeclaration = file.getInterface(referenceType.name);
    if (interfaceDeclaration) {
      const name = interfaceDeclaration.getName();
      const description = extractJsDocComment(interfaceDeclaration);
      const type = parseInterfaceDeclaration(interfaceDeclaration);
      return { description, name, type };
    }

    throw new Error("unexpected type reference");
  });

  return { api, endpoints, types };
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics.map(diagnostic => diagnostic.getMessageText()).join("\n")
    );
  }
}
