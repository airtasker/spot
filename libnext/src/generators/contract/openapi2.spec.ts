import * as path from "path";
import { cleanse } from "../../cleansers/cleanser";
import { parseFilePath } from "../../parsers/parser";
import { generateOpenApiV2 } from "./openapi2";

const EXAMPLE_PATH = path.join(
  __dirname,
  "..",
  "..",
  "test",
  "examples",
  "contract.ts"
);

describe("OpenAPI 2 generator", () => {
  test("produces valid code", async () => {
    const contractNode = await parseFilePath(EXAMPLE_PATH, {
      baseUrl: ".",
      paths: {
        "@airtasker/spotnext": ["./libnext/src/lib"]
      }
    });
    const contractDefinition = cleanse(contractNode);
    expect(generateOpenApiV2(contractDefinition, "json")).toMatchSnapshot(
      "json"
    );
    expect(generateOpenApiV2(contractDefinition, "yaml")).toMatchSnapshot(
      "yaml"
    );
  });
});
