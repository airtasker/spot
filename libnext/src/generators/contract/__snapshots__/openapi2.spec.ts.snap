// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI 2 generator produces valid code: json 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"version\\": \\"0.0.0\\",
    \\"title\\": \\"company-api\\",
    \\"description\\": \\"This is the company API. It does cool things\\",
    \\"contact\\": {
      \\"name\\": \\"TODO\\"
    }
  },
  \\"paths\\": {
    \\"/company/{companyId}/users/{userId}\\": {
      \\"post\\": {
        \\"operationId\\": \\"GetUser\\",
        \\"description\\": \\"Retrieves a user in a company\\",
        \\"tags\\": [
          \\"Company\\",
          \\"User\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"companyId\\",
            \\"description\\": \\"company identifier\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          },
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"userId\\",
            \\"description\\": \\"user identifier\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          },
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"sample-query\\",
            \\"description\\": \\"a demo query param\\",
            \\"type\\": \\"string\\",
            \\"required\\": false
          },
          {
            \\"in\\": \\"header\\",
            \\"name\\": \\"x-auth-token\\",
            \\"description\\": \\"Auth Header\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"201\\": {
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/UserBody\\"
            },
            \\"description\\": \\"Successful creation of user\\"
          },
          \\"404\\": {
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/ErrorBody\\"
            },
            \\"description\\": \\"Bad request response\\"
          }
        }
      }
    },
    \\"/company/{companyId}/users\\": {
      \\"post\\": {
        \\"operationId\\": \\"CreateUser\\",
        \\"description\\": \\"Creates a user in a company\\",
        \\"tags\\": [
          \\"Company\\",
          \\"User\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"companyId\\",
            \\"description\\": \\"company identifier\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          },
          {
            \\"in\\": \\"body\\",
            \\"name\\": \\"body\\",
            \\"description\\": \\"TODO\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/CreateUserRequestBody\\"
            }
          },
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"sample-query\\",
            \\"description\\": \\"a demo query param\\",
            \\"type\\": \\"string\\",
            \\"required\\": false
          },
          {
            \\"in\\": \\"header\\",
            \\"name\\": \\"x-auth-token\\",
            \\"description\\": \\"Auth Header\\",
            \\"type\\": \\"string\\",
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"201\\": {
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/UserBody\\"
            },
            \\"description\\": \\"Successful creation of user\\"
          },
          \\"400\\": {
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/ErrorBody\\"
            },
            \\"description\\": \\"Bad request response\\"
          },
          \\"default\\": {
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/ErrorBody\\"
            },
            \\"description\\": \\"\\"
          }
        }
      }
    }
  },
  \\"definitions\\": {
    \\"UserBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"data\\": {
          \\"type\\": \\"object\\",
          \\"properties\\": {
            \\"firstName\\": {
              \\"type\\": \\"string\\"
            },
            \\"lastName\\": {
              \\"type\\": \\"string\\"
            },
            \\"profile\\": {
              \\"$ref\\": \\"#/definitions/Profile\\"
            }
          },
          \\"required\\": [
            \\"firstName\\",
            \\"lastName\\",
            \\"profile\\"
          ]
        }
      },
      \\"required\\": [
        \\"data\\"
      ]
    },
    \\"ErrorBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        },
        \\"message\\": {
          \\"type\\": \\"array\\",
          \\"items\\": {
            \\"type\\": \\"string\\"
          }
        }
      },
      \\"required\\": [
        \\"name\\",
        \\"message\\"
      ]
    },
    \\"CreateUserRequestBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"data\\": {
          \\"type\\": \\"object\\",
          \\"properties\\": {
            \\"firstName\\": {
              \\"type\\": \\"string\\"
            },
            \\"lastName\\": {
              \\"type\\": \\"string\\"
            },
            \\"age\\": {
              \\"type\\": \\"number\\"
            },
            \\"email\\": {
              \\"$ref\\": \\"#/definitions/Email\\"
            },
            \\"address\\": {
              \\"$ref\\": \\"#/definitions/Address\\"
            }
          },
          \\"required\\": [
            \\"firstName\\",
            \\"lastName\\",
            \\"age\\",
            \\"email\\",
            \\"address\\"
          ]
        }
      },
      \\"required\\": [
        \\"data\\"
      ]
    },
    \\"Profile\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"private\\": {
          \\"type\\": \\"boolean\\"
        },
        \\"messageOptions\\": {
          \\"$ref\\": \\"#/definitions/MessageOptions\\"
        }
      },
      \\"required\\": [
        \\"private\\",
        \\"messageOptions\\"
      ]
    },
    \\"MessageOptions\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"newsletter\\": {
          \\"type\\": \\"boolean\\"
        }
      },
      \\"required\\": [
        \\"newsletter\\"
      ]
    },
    \\"Email\\": {
      \\"type\\": \\"string\\"
    },
    \\"Address\\": {
      \\"type\\": \\"string\\"
    }
  }
}"
`;

exports[`OpenAPI 2 generator produces valid code: yaml 1`] = `
"swagger: '2.0'
info:
  version: 0.0.0
  title: company-api
  description: This is the company API. It does cool things
  contact:
    name: TODO
paths:
  '/company/{companyId}/users/{userId}':
    post:
      operationId: GetUser
      description: Retrieves a user in a company
      tags:
        - Company
        - User
      parameters:
        - in: path
          name: companyId
          description: company identifier
          type: string
          required: true
        - in: path
          name: userId
          description: user identifier
          type: string
          required: true
        - in: query
          name: sample-query
          description: a demo query param
          type: string
          required: false
        - in: header
          name: x-auth-token
          description: Auth Header
          type: string
          required: true
      responses:
        '201':
          schema:
            $ref: '#/definitions/UserBody'
          description: Successful creation of user
        '404':
          schema:
            $ref: '#/definitions/ErrorBody'
          description: Bad request response
  '/company/{companyId}/users':
    post:
      operationId: CreateUser
      description: Creates a user in a company
      tags:
        - Company
        - User
      parameters:
        - in: path
          name: companyId
          description: company identifier
          type: string
          required: true
        - in: body
          name: body
          description: TODO
          required: true
          schema:
            $ref: '#/definitions/CreateUserRequestBody'
        - in: query
          name: sample-query
          description: a demo query param
          type: string
          required: false
        - in: header
          name: x-auth-token
          description: Auth Header
          type: string
          required: true
      responses:
        '201':
          schema:
            $ref: '#/definitions/UserBody'
          description: Successful creation of user
        '400':
          schema:
            $ref: '#/definitions/ErrorBody'
          description: Bad request response
        default:
          schema:
            $ref: '#/definitions/ErrorBody'
          description: ''
definitions:
  UserBody:
    type: object
    properties:
      data:
        type: object
        properties:
          firstName:
            type: string
          lastName:
            type: string
          profile:
            $ref: '#/definitions/Profile'
        required:
          - firstName
          - lastName
          - profile
    required:
      - data
  ErrorBody:
    type: object
    properties:
      name:
        type: string
      message:
        type: array
        items:
          type: string
    required:
      - name
      - message
  CreateUserRequestBody:
    type: object
    properties:
      data:
        type: object
        properties:
          firstName:
            type: string
          lastName:
            type: string
          age:
            type: number
          email:
            $ref: '#/definitions/Email'
          address:
            $ref: '#/definitions/Address'
        required:
          - firstName
          - lastName
          - age
          - email
          - address
    required:
      - data
  Profile:
    type: object
    properties:
      private:
        type: boolean
      messageOptions:
        $ref: '#/definitions/MessageOptions'
    required:
      - private
      - messageOptions
  MessageOptions:
    type: object
    properties:
      newsletter:
        type: boolean
    required:
      - newsletter
  Email:
    type: string
  Address:
    type: string
"
`;
